ENH: add edge strength plotting to DAG.to_daft() and DAG.to_graphviz

Dhiren-Mhatre opened the pull request:
## Add Edge Strength Plotting to DAG Visualization Methods

This pull request implements GitHub issue #2146 by adding edge strength plotting functionality to pgmpy's DAG class visualization methods.

### Description

Add `plot_edge_strength` parameter to both `DAG.to_daft()` and `DAG.to_graphviz()` methods that displays edge strength values as labels when enabled. The implementation includes comprehensive testing, maintains backward compatibility, and follows pgmpy coding standards.

### Your checklist for this pull request
Please review the [guidelines for contributing](https://pgmpy.org/started/contributing.html) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2146

### List of changes to the codebase in this pull request
- **Enhanced `DAG.to_daft()` method**: Added `plot_edge_strength=False` parameter that displays edge strength values as labels on edges when enabled
- **Enhanced `DAG.to_graphviz()` method**: Added `plot_edge_strength=False` parameter with same functionality for pygraphviz backend  
- **Edge strength formatting**: Values are displayed formatted to 3 decimal places for consistent presentation
- **Integration with existing systems**: Works seamlessly with existing `edge_params` and user-provided edge labels
- **Error handling**: Graceful handling of missing dependencies (daft/pygraphviz) and missing edge strength data with appropriate warning logging
- **Comprehensive testing**: Added 15+ pytest tests covering API acceptance, functional behavior, missing dependencies, edge cases, and integration scenarios
- **Backward compatibility**: Default behavior preserved (`plot_edge_strength=False`) ensuring existing code remains unaffected
- **Documentation**: Updated docstrings with parameter descriptions, usage examples, and requirements
- **Code quality**: Follows pgmpy standards with Black formatting, proper error handling, and descriptive variable names


### Testing

All tests pass with comprehensive coverage including:
- API parameter acceptance tests (no optional dependencies required)
- Functional tests with edge strength display
- Missing dependency handling tests  
- Integration tests with existing edge parameters
- Error condition and edge case handling

The implementation maintains full backward compatibility while adding the requested edge strength visualization capability as specified in issue #2146.

Dhiren-Mhatre committed:
feat: add edge strength plotting to DAG.to_daft() and DAG.to_graphviz() - Add plot_edge_strength parameter to both visualization methods - Displays edge strength values as labels when enabled - Includes comprehensive testing and backward compatibility - Closes #2146

Dhiren-Mhatre committed:
test: improve edge strength plotting test coverage

Dhiren-Mhatre committed:
Fix codecov coverage issues in DAG tests

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -21,6 +21,21 @@
 from pgmpy.models import LinearGaussianBayesianNetwork as LGBN
 
+# Check for optional packages
+try:
```
All optional packages are installed for testing environment, so this is not needed

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -678,6 +693,50 @@ def test_edge_strength_skip_latent_edges(self):
+        self.assertEqual(dag.edges[("A", "B")]["strength"], 0.123)
+        self.assertEqual(dag.edges[("C", "B")]["strength"], 0.456)
+
+        # Test edge strength formatting
```
What exactly is this testing? Rounding?

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -678,6 +693,50 @@ def test_edge_strength_skip_latent_edges(self):
+            formatted = f"{value:.3f}"
+            self.assertEqual(formatted, expected_formatted[i])
+
+    def test_optional_package_imports(self):
```
This needs to be removed.

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -678,6 +693,50 @@ def test_edge_strength_skip_latent_edges(self):
+        for i, value in enumerate(test_values):
+            formatted = f"{value:.3f}"
+            self.assertEqual(formatted, expected_formatted[i])
+
```
We need to test whether the objects returned by the two methods has the correct edge label assigned.

ankurankan reviewed (changes requested):
@Dhiren-Mhatre Thanks for the PR. Could you also share how these plots look like?

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1228,6 +1233,22 @@ def to_daft(
                 extra_params = dict()
+
+            # Add edge strength as label if requested
+            if plot_edge_strength:
```
Need to add a check to first check if the `edge_strengths` exist.

ankurankan replied:
This needs to be at the start of the method. We want to check all the arguments at the beginning so that the method's logic is cleaner.

Dhiren-Mhatre committed:
applied suggestions

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

Dhiren-Mhatre committed:
Added Comprehensive Test Coverage for missing lines in DAG.py

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

Dhiren-Mhatre committed:
fixed a failing test

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

Dhiren-Mhatre commented:
@ankurankan Here are examples of how the edge strength plotting looks:
![edge_strength_demo](https://github.com/user-attachments/assets/8db39d5f-7f1e-45b3-a63c-31035fb4c0ac)

I've now addressed all of your feedback points:
 Removed optional package import checks  but CI  is failing with reason  ModuleNotFoundError: No module named 'pygraphviz', any thoughts on how we should handle that dependency?
  Removed the edge strength formatting test as it only tested Python's built-in formatting . 
  Removed the optional package imports test method.
  Moved validation logic to the beginning of both `to_daft()` and `to_graphviz()` methods for cleaner code structure.
  Enhanced edge strength existence check with proper error handling.

ankurankan commented:
@Dhiren-Mhatre For the tests, you will need to add pygraphviz to `pyproject.toml`. It should go in test dependencies.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1228,6 +1246,22 @@ def to_daft(
+
+            # Add edge strength as label if requested
+            if plot_edge_strength:
+                if "strength" in self.edges[(u, v)]:
```
@Dhiren-Mhatre Why are these checks repeated? After the first check, the values are guaranteed to exist. Right?

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1163,6 +1168,19 @@ def to_daft(
+                    missing_strengths.append((u, v))
+
+            if missing_strengths:
+                logger.warning(
```
We should throw an error here instead of a warning.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1301,20 +1335,56 @@ def get_random(
+                    missing_strengths.append((u, v))
+
+            if missing_strengths:
+                logger.warning(
```
Same, we should throw an error here.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1301,20 +1335,56 @@ def get_random(
+        # Add edge strength labels if requested
+        if plot_edge_strength:
+            for u, v in self.edges():
+                if "strength" in self.edges[(u, v)]:
```
If we throw error above, no need to do the checks again.

Dhiren-Mhatre committed:
Address reviewer comments: replace warnings with errors, remove redundant checks, add pygraphviz dependency

Dhiren-Mhatre commented:
> @Dhiren-Mhatre For the tests, you will need to add pygraphviz to `pyproject.toml`. It should go in test dependencies.

 I’ve added pygraphviz to the test dependencies in pyproject.toml. Appreciate the guidance!

ankurankan reviewed (changes requested):
Hmm, so pip can't install the graphviz package which is required by pygraphviz. This would be complicated to solve on github actions because of different platforms. But I don't think we need it for tests. I see that you import `pygraphviz` in the test file, but it doesn't seem to be getting used anywhere. Am I missing it somewhere?

ankurankan commented on pyproject.toml:
```diff
@@ -47,7 +47,8 @@ tests = [
   "black",
-  "pre-commit"
+  "pre-commit",
+  "pygraphviz"
```
Sorry, this should be in optional not in tests.

Dhiren-Mhatre committed:
changed location of pygraphviz from tests to optional

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

Dhiren-Mhatre commented:
You're right , I have moved pygraphviz to the optional dependencies as requested. However, pygraphviz is being used extensively in the tests like in  `ab_edge = graphviz_plot.get_edge("A", "B")  # pygraphviz method`

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1228,6 +1246,16 @@ def to_daft(
+            # Add edge strength as label if requested
+            if plot_edge_strength:
+                strength_value = self.edges[(u, v)]["strength"]
+                # Format the strength value to 3 decimal places
```
I don't think we need comments explaining every line.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1228,6 +1246,16 @@ def to_daft(
+                # Format the strength value to 3 decimal places
+                strength_label = f"{strength_value:.3f}"
+                # If user didn't provide a custom label, use the strength
+                if "label" not in extra_params:
```
This will need error handling. Right? If label is already in `extra_params`

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -482,23 +484,24 @@ def test_edge_strength_basic(self):
+        self.assertIn(("Z", "Y"), strengths)
 
-        # Test that edge strengths match squared Pearson correlation
-        xy_corr = pearsonr("X", "Y", ["Z"], data, boolean=False)
```
Why was this test removed?

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -678,6 +681,346 @@ def test_edge_strength_skip_latent_edges(self):
+        with self.assertRaises(ValueError):
+            dag.to_graphviz(plot_edge_strength=True)
+
+    def test_edge_strength_plotting_format_precision(self):
```
Don't really need this. This is a lot of tests for a relatively simple functionality. This would make maintenance harder in the long term.

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -678,6 +681,346 @@ def test_edge_strength_skip_latent_edges(self):
+        # Test all code paths in edge strength plotting methods
+
+        # Scenario 1: Mixed edge strengths for to_daft individual errors
+        dag1 = DAG([("A", "B"), ("C", "B")])
```
Isn't this repeated test?

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -678,6 +681,346 @@ def test_edge_strength_skip_latent_edges(self):
+        graphviz_plot3 = dag3.to_graphviz(plot_edge_strength=True)
+        self.assertIsNotNone(graphviz_plot3)
+
+    def test_optional_dependency_coverage(self):
```
Don't need this.

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -678,6 +681,346 @@ def test_edge_strength_skip_latent_edges(self):
+        except Exception as e:
+            self.fail(f"to_graphviz failed when pygraphviz is available: {e}")
+
+    def test_get_random_comprehensive(self):
```
Isn't `get_random` method already has tests?

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -1012,16 +1355,6 @@ def test_has_undirected_edge(self):
         self.assertTrue(pdag.has_undirected_edge("A", "B"))
         self.assertTrue(pdag.has_undirected_edge("B", "A"))
 
-    def test_undirected_neighbors(self):
```
Why was this removed?

ankurankan reviewed (changes requested):
I am assuming you have generated tests using AI. That is way too much additional code for a relatively simple functionality. You will need to compress some of the tests. You have also removed some very important tests, please be more careful about these.

> You're right , I have moved pygraphviz to the optional dependencies as requested. However, pygraphviz is being used extensively in the tests like in ab_edge = graphviz_plot.get_edge("A", "B")  # pygraphviz method

Yes, but you don't explicitly call any pygraphviz function so you don't need to import it. Tests will work if you just remove the import statement.

Dhiren-Mhatre committed:
applied suggestions

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

Dhiren-Mhatre committed:
ENH: add edge strength plotting to DAG.to_daft() and DAG.to_graphviz

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

Dhiren-Mhatre committed:
ENH: add edge strength plotting to DAG.to_daft() and DAG.to_graphviz

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

Dhiren-Mhatre committed:
Revert "ENH: add edge strength plotting to DAG.to_daft() and DAG.to_graphviz"

This reverts commit f01c75197218e5560f547a3c3501fc83f5788101.

ankurankan commented:
@Dhiren-Mhatre Can you add the following in `ci.yml` before the "Print python info" line:
```
      - name: Install graphviz (Ubuntu)
        if: matrix.os == 'ubuntu-latest'
        run: sudo apt-get update && sudo apt-get install graphviz -y

      - name: Install graphviz (Windows)
        if: matrix.os == 'windows-latest'
        run: choco install graphviz -y

      - name: Install graphviz (macOS)
        if: matrix.os == 'macOS-latest'
        run: brew install graphviz
```
We will need graphviz for the CI to pass.

Dhiren-Mhatre committed:
updated ci.yml to add graphviz

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

ankurankan commented:
@Dhiren-Mhatre: Tests are failing. Can you please fix them?

Dhiren-Mhatre committed:
fixed tests

Signed-off-by: Dhiren-Mhatre <kp064669@gmail.com>

ankurankan committed:
Merge branch 'dev' into feature/edge-strength-plotting-issue-2146

ankurankan committed:
Fix black and flake8

ankurankan committed:
Minor change

ankurankan commented:
@Dhiren-Mhatre I have pushed some formatting fixes. Please install the pre-commit hook if you haven't yet. The instructions are in the Contributing guide: https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md

ankurankan committed:
Fixes tests

ankurankan commented:
@Dhiren-Mhatre Thanks for the PR. Looks good to me, I have added a few minor commits.

Dhiren-Mhatre commented:
> @Dhiren-Mhatre Thanks for the PR. Looks good to me, I have added a few minor commits.

Thank you! I’ll make sure to apply what I’ve learned from this PR to future contributions.
